OpenTerm/Util/Cub.swift

```diff
@@ -0,0 +1,160 @@
+//
+//  Cub.swift
+//  OpenTerm
+//
+//  Created by Louis D'hauwe on 03/02/2018.
+//  Copyright © 2018 Silver Fox. All rights reserved.
+//
+
+import Foundation
+import Cub
+import ios_system
+import TabView
+
+public func cub(argc: Int32, argv: UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>?) -> Int32 {
+
+	let tabViewContainer = UIApplication.shared.keyWindow?.rootViewController as! TabViewContainerViewController<TerminalTabViewController>
+
+	guard let activeVC = tabViewContainer.primaryTabViewController.visibleViewController as? TerminalViewController else {
+		return 0
+	}
+
+	activeVC.terminalView.isExecutingScript = true
+	let executor = activeVC.terminalView.executor
+
+	guard argc == 2 else {
+		fputs("Usage: cub script.cub\n", thread_stderr)
+		return 1
+	}
+
+	guard let fileName = argv?[1] else {
+		fputs("Usage: cub script.cub\n", thread_stderr)
+		return 1
+	}
+
+	let path = String(cString: fileName)
+
+	guard FileManager.default.fileExists(atPath: path) else {
+		fputs("Missing file \(path)\n", thread_stderr)
+		return 1
+	}
+
+	let url = URL(fileURLWithPath: path)
+
+	guard let data = FileManager.default.contents(atPath: url.path) else {
+		fputs("Missing file \(path)\n", thread_stderr)
+		return 1
+	}
+
+	guard let source = String.init(data: data, encoding: .utf8) else {
+		fputs("Missing file \(path)\n", thread_stderr)
+		return 1
+	}
+
+	let runner = Runner(logDebug: true, logTime: false)
+
+	runner.registerExternalFunction(name: "print", argumentNames: ["input"], returns: false) { (arguments, callback) in
+
+		for (name, arg) in arguments {
+			fputs("\(arg)\n", thread_stdout)
+		}
+
+		callback(nil)
+		return
+	}
+
+	runner.registerExternalFunction(name: "exec", argumentNames: ["command"], returns: true) { (arguments, callback) in
+
+		var arguments = arguments
+
+		guard let command = arguments.removeValue(forKey: "command") else {
+			callback(.number(1))
+			return
+		}
+
+		guard case let .string(commandStr) = command else {
+			callback(.number(1))
+			return
+		}
+
+		print("run command: \(commandStr)")
+
+		//        executor.dispatch(commandStr, callback: { (code) in
```

Comment:
What will you end up doing here for execution? Will you need to add a callback?

Reply:
I've added CubCommandExecutor, which conforms to CommandExecutorDelegate and ParserDelegate. It takes a TerminalView and temporarily sets itself as the delegates.